flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,230 @@
+                .values()
+                .forEach(
+                        changeSet -> {
+                            if (changeSet.isConfirmed()) {
+                                result.toReturn.add(changeSet);
+                            } else if (changeSet.setScheduled()) {
+                                result.toUpload.add(changeSet);
+                            } else {
+                                // we also re-upload any scheduled/uploading/uploaded changes
+                                // even if they were not sent to the JM yet because this can happen
+                                // in the meantime and then JM can decide to discard them
+                                result.toRetry.add(changeSet.forRetry());
+                            }
+                        });
+        result.toUpload.addAll(result.toRetry);
+        result.toReturn.addAll(result.toUpload);
+        return result;
+    }
+
+    @Override
+    public void close() {
+        LOG.debug("close {}", logId);
+        checkState(!closed);
+        closed = true;
+        activeChangeSet.clear();
+        changeSets.values().forEach(StateChangeSet::setCancelled);
+        // todo in MVP or later: cleanup if transition succeeded and had non-shared state
+        changeSets.clear();
+        // the store is closed from the owning FsStateChangelogClient
+    }
+
+    @Override
+    public void confirm(SequenceNumber from, SequenceNumber to) {
+        LOG.debug("confirm {} from {} to {}", logId, from, to);
+        changeSets.subMap(from, true, to, false).values().forEach(StateChangeSet::setConfirmed);
+    }
+
+    @Override
+    public void reset(SequenceNumber from, SequenceNumber to) {
+        LOG.debug("reset {} from {} to {}", logId, from, to);
+        changeSets.subMap(from, to).forEach((key, value) -> value.setAborted());
+        // todo in MVP or later: cleanup if change to aborted succeeded and had non-shared state
+        // For now, relying on manual cleanup.
+        // If a checkpoint that is aborted uses the changes uploaded for another checkpoint
+        // which was completed on JM but not confirmed to this TM
+        // then discarding those changes would invalidate that previously completed checkpoint.
+        // Solution:
+        // 1. pass last completed checkpoint id in barriers, trigger RPC, and abort RPC
+        // 2. confirm for the id above
+        // 3. make sure that at most 1 checkpoint in flight (CC config)
+    }
+
+    @Override
+    public void truncate(SequenceNumber to) {
+        LOG.debug("truncate {} to {}", logId, to);
+        if (to.compareTo(lastAppendedSequenceNumber) > 0) {
+            // can happen if client calls truncate(prevSqn.next())
+            rollover();
+        }
+        NavigableMap<SequenceNumber, StateChangeSet> headMap = changeSets.headMap(to, false);
+        headMap.values().forEach(StateChangeSet::setTruncated);
+        headMap.clear();
+    }
```

Comment:
Are we holding in memory all state changes until we call truncate?

Reply:
The changes themselves (byte arrays) are kept until JM comfirmation or this truncate call.
ChangeSet objects are kept until this call.